src/constants.rs

```diff
@@ -0,0 +1,321 @@
+// Keep this sorted!
+//
+// whitelist fetched from http://spdx.org/licenses/
+
+pub const DOCUMENT_REFERENCE: &'static str = "^(DocumentRef-[0-9A-Za-z.-]+:)?LicenseRef-[0-9A-Za-z.-]+$";
```

Comment:
Hrmm. This crate claims to be an implementation of the SPDX spec, so it should conform to it. However, the main (only?) client of this library is crates.io, which has its own convention for using a license not on this list.

Also, the SPDX spec supports two other options for a license - "NONE" (there is no license) and "NOASSERTION" (unknown license).

Basically, I think we need to parse the Ok value into an actual information carrying value instead of `()`, that way crates.io can avoid accepting document references, NONEs, and NOASSERTIONS. A simple enum like this is fine for our purposes:

```rust
enum License<'a> {
    Expression(&'a str),
    Reference(&'a str),
    None,
    NoAssertion,
}
```

A more complete implementation would return a fully parsed tree for the expression value, but I'm not sure its worth the effort. The only argument for doing that I can think of is that that would be a breaking change and this is a breaking change now.

What we can do is keep `validate_license_expr`'s signature the same, and have a new function `parse_license_expr` that returns the fuller value.

Reply:
> Also, the SPDX spec supports two other options for a license - "NONE" (there is no license) and "NOASSERTION" (unknown license).

While these are currently options [for SPDX license declarations][1], they are not part of the [license expression syntax][2] (although I would like to see them added, spdx/spdx-spec#49, spdx/spdx-spec#50).  But `LicenseRef-` are part of the expression syntax, so your point stands.  If the goal is to disallow document references (and eventually, if they become legal, `NONE` and `NOASSERTION`) in crates.io, then it would make sense to have `validate_license_expr` take options to block those at the caller's preference.

[1]: https://spdx.org/spdx-specification-21-web-version#h.ihv636
[2]: https://spdx.org/spdx-specification-21-web-version#h.jxpfx0ykyb60

Reply:
On the validation front, it would also be good to have toggle or result that could warn on deprecated constructs or identifiers.  More on this [here][1] and [here][2].

[1]: https://github.com/rust-lang/cargo/issues/2039#issuecomment-236644901
[2]: https://github.com/withoutboats/license-exprs/pull/6#issuecomment-232207918